index.html

```diff
@@ -245,7 +249,9 @@ <h3>
                 <a href="https://w3c.github.io/vc-data-model/#json-web-token">VC-JWT</a>,
                 <a href="https://w3c-ccg.github.io/lds-ed25519-2020/">Ed25519 Cryptosuite</a>,
                 <a href="https://w3c-ccg.github.io/lds-ecdsa-secp256k1-2019/">Secp256k1 Cryptosuite</a>,
-                <a href="https://w3c-ccg.github.io/ldp-bbs2020/">BBS+ Cryptosuite</a>
+                <a href="https://w3c-ccg.github.io/ldp-bbs2020/">BBS+ Cryptosuite</a>,
+                <a href="https://ns.did.ai/suites/multikey-2021/">Multikey-2021</a>
```

Comment:
-1 for me

I don't buy into the idea that the WG should be defining key formats in the VCWG and that should remain separately abstracted from the signature formats.

Reply:
because of the choices made by the CCG, the key and signature formats are defined in the same specs. So if you want to include an input document, you are forced to refer to both.

There is also a community perspective that each key representation should be used for 1 signature representation.

This aligns with how P-384 is used with ES384 in JOSE.

I am inclined to agree that it would be better to just leverage JOSE / COSE key and signature types.... this would lead to a larger supply of software providers and better adoptions by legacy systems.

Reply:
> There is also a community perspective that each key representation should be used for 1 signature representation.

I think this is likely to be overridden in the VC2WG... but again, that's a discussion that we have to have in that group, and in order for us to have that discussion, all perspectives must be inputs into the VC2WG. So, @kdenhartog -- we need to clearly signal that a subset of the group does intend to define the Multikey format. I don't think we need to refer to Orie's Multikey-2021 JSON-LD Context for that since we very clearly call it out already in one of the input documents already:

https://w3c-ccg.github.io/data-integrity-spec/#multikey

... and in other cryptosuite specs that are input documents:

https://w3c-ccg.github.io/di-ecdsa-secpr1-2019/#example-a-multibase-encoded-multicodec-curve-p-384-public-key

https://w3c-ccg.github.io/di-ecdsa-secpr1-2019/#example-a-dataintegritysignature-example-using-a-nistecdsa2019-cryptosuite

The text clearly still needs work, but there is an expectation written in these input documents that we do plan to separate key format from cryptosuite in the general case, but still provide the option to specify key format in cryptosuites to support things like EthereumEip712Signature2020 (which doesn't have a JWK/JWS mapping, nor a Multikey mapping, and probably won't because it's more proof than signature). PGP falls into the same category of not having mappings in either JWK/JWS or Multikey. PEM, while having a JWK mapping (kinda sorta, but not really because many implementations do totally different things there) also falls into this category.

So, in other words @kdenhartog -- we are going to have to define some key formats because JWK doesn't work for everything, and Multikey has clear benefits over JWK (like the ability to express a public key in a very simple an compact manner -- did:key identifiers **are** public Multikeys.

Reply:
> that's a discussion that we have to have in that group, and in order for us to have that discussion, all perspectives must be inputs into the VC2WG.

This is the strongest argument for why I think this should be in here, but because we accept the work as a possible path in the charter it doesn't mean it will be a required path to take once we get into the work.

Generally speaking, I've found that by separating the key format from the proof format it becomes easier to rely on a variety of different forms of PKI infrastructure (PGP servers, DIDs, x509 certs, .well-known JWK sets, etc) in a more interoperable way. As long as the public key contains the proper associated metadata (curve/algo/key parameters/usage constraints/etc) the way in which the key material and metadata is formatted is largely irrelevant in my opinion since that becomes an implementation concern. This is because it's relatively easy to convert from one format to another during verification and that's why I'm arguing that key formats should be left out of scope. In this body of work it's my belief that we should be focusing on the proof formats and leave the key formats to the already defined formats that we're done elsewhere including DIDs, x509, JWK sets, and PEM key formats.

Reply:
> it's relatively easy to convert from one format to another during verification

It might be easy for you personally (@kdenhartog) to convert from every key format to every other key format, but it is not easy for the average developer. I certainly wouldn't trust myself to write that code without some really great test suites and even then, I'd need lots and lots of code review on the library to even remotely feel safe pushing it out to production. 

That is, it seems simple enough to slap together some code for a particular use case, but it's no simple feat to take a JWK encoded RSA public key and transform it, properly, into something that a FIPS certified OpenSSL library will use... much less /every key format/ in significant use.

I mean, can you point me to the PEM -> JWK transformation conformance test suite? What about any of the JOSE conformance test suites, can you point me to those? The x509 PEM test suite? I really want you to try to find these things. You might discover that they don't exist... and that all the care that you think has been put into the safety of the cryptography libraries that you trust just might not actually be there.

If we are going to suggest that people do arbitrary key transformations, and that should be out of scope for the WG... we should at least make sure that someone else has the ball on that (I don't think anyone has the ball, there).

> it's my belief that we should be focusing on the proof formats and leave the key formats to the already defined formats

What you're saying above seems to be: "I don't think that multikey should exist." ... and if you believe that, then that raises the question on whether or not you feel like did:key should exist. :)

So, I'm getting increasingly confused about your mental model wrt. key transformation and multikey. Please help me understand your mental model.

Reply:
>It might be easy for you personally (@kdenhartog) to convert from every key format to every other key format, but it is not easy for the average developer.

The average developer is able to figure out how to decode bytes and re-encode bytes to another format especially in the well defined formats that are commonly used throughout this ecosystem. Which of the aforementioned format conversions is doing something greater than encoding and translating one data model into another? It seems like you're interpreting what I'm saying to mean that in order to format a key into a particular data model that it should also be validated which doesn't make sense because the domain parameters of the key material are tied to the algorithm their used with. This is why the domain parameters in NIST documents are defined relative to the algorithm rather than the curve in use. Therefore, _key validation_ is expected to be done when the algorithm is executed. Not when the key material is converted from one data model/encoding to another. So where is the confusion coming from here?

>What you're saying above seems to be: "I don't think that multikey should exist."

That's a strawman argument that I've not made. However an argument I would make is that multikey is poorly defined at the moment because it has built in footguns by not including any additional key metadata to specify the usage of the key. By not including metadata about the algorithm the key is expected to be used for it may lead to developers reusing the same key material for a variety of different algorithms like did:key endorses today through assigning verification relationships for usage of ECDH and signature algorithms by default. If this was done with an [RSA key](https://github.com/multiformats/multicodec/blob/master/table.csv#L147) where it's being used for multiple algorithms then it's susceptible to a variety of different attacks. This would be prevented if the key format encoded the additional metadata about the key and not only the key material.

> Please help me understand your mental model.

I'm focused on improving interoperability through a different abstraction than what you're advocating for. It's my belief that by not tying the key format to the signature format we're enabling the VC data model to interoperate across a variety of different public key infrastructures already in use today rather than to push for a "not invented here" model that ties the data model to a predefined set of PKI systems.

Reply:
@msporny since this is closed I'm not sure if you're still getting notified for this thread. Just want to ping you here again specifically since I forgot to a few days ago to respond to the comments and provide better reasoning as to why I'm taking the position I am.

Reply:
> It's my position that by not tying the key format to the signature format we're enabling the VC data model to interoperate across a variety of different public key infrastructures already in use today rather than to push for a "must be coupled" model that ties the data model to a predefined set of PKI systems.

The above text shows that we're talking past each other. :) 

I expect the group to push towards your position here, and I can see the advantages of doing so. I won't say "there are zero problems with doing that" at present because we haven't been able to see what the security attack surface looks like if we fully decouple keys from cryptosuites. If we are going to allow arbitrary key translation, we need to have a very thorough test suite for doing that e.g., (PEM, JWK, Multikey, COSE) -> (PEM, JWK, Multikey, COSE) -- that's 12 code paths to convert from one to another that need to be tested for 46 key types listed in the JOSE registry alone, so... 552 tests (at a minimum). 

It's do-able, but we're talking about requiring some pretty serious effort on that test suite alone. If you're volunteering to write that test suite and drive implementers to implement, that would go a long way to alleviate my concerns around arbitrary key translation.

Reply:
I believe you're making the assumption that this is going to be normatively defined hence the need for the testing. My assumption is that this is an implementation detail depending on the PKI system that's chosen to be supported. This aligns with the thinking used by Web Crypto where the internal data model for storing keys is different from the formats that are supported in the importKey/exportKey APIs.

Reply:
> Never leave critical security infrastructure to "an implementation detail". I know that tends to be a standard practice in today's security industry, but that's wrong, and we should start calling it out.

This is literally what your code that I referenced that _you_ and @dmitrizagidulin wrote is doing though. How are you arguing for this and then implementing what I'm arguing for?

Reply:
> This is literally what your code that I referenced that _you_ and @dmitrizagidulin wrote is doing though. How are you arguing for this and then implementing what I'm arguing for and then suggesting what I'm proposing is wrong? It seems outright contradictory to me.

Because what we (myself and @dmitrizagidulin) are doing, right now, is completely and utterly wrong. Perhaps that's the miscommunication. I think what everyone is doing right now wrt. hand-waving over key translation being an implementation detail, including us, is wrong.

If we are to tell the world that they should depend on key translation, we need a test suite that ensures everyone is doing this correctly. That needs to be fixed, for everyone in the ecosystem.

It's not contradictory, I'm saying all of us, everyone, including myself and @dmitrizagidulin, cannot keep doing what we're doing now, which is doing key translation w/o a community approved test suite to ensure that what we're doing is correct. As an industry, we have to do better. :)

Does that help?

Reply:
I think we fundamentally just see things differently than because I don't believe it's wrong, in fact I think it's the better abstraction model. With that I think we're going to need to figure out what the best way to support both sides because I'm adamant that standardizing key representations are unnecessary, the authorization model is not understood to be a authorization system (there's probably 10 or 20 of us total who understand that verification relationships are an authz system), and the things you're advocating for will be ignored as they already have been and as I've watched numerous new developers do when watching how they use DIDs and VCs.